File: data/derived_cot/rq1_traces/translation/qwen/java_to_py/code_output/Snake.py
import random

class Snake:
    class Position:
        def __init__(self, x, y):
            self.x = x
            self.y = y
        
        def __eq__(self, other):
            if isinstance(other, Snake.Position):
                return self.x == other.x and self.y == other.y
            return False
        
        def __hash__(self):
            return hash((self.x, self.y))
        
        def getX(self):
            return self.x
        
        def getY(self):
            return self.y
    
    def __init__(self, SCREEN_WIDTH, SCREEN_HEIGHT, BLOCK_SIZE, food_position):
        self.length = 1
        self.SCREEN_WIDTH = SCREEN_WIDTH
        self.SCREEN_HEIGHT = SCREEN_HEIGHT
        self.BLOCK_SIZE = BLOCK_SIZE
        self.positions = []
        self.positions.append(Snake.Position(SCREEN_WIDTH // 2, SCREEN_HEIGHT // 2))
        self.score = 0
        self.food_position = food_position
        self.random = random.Random()
    
    def move(self, direction):
        cur = self.positions[0]
        x = direction.getX()
        y = direction.getY()
        
        new_x = (cur.x + (x * self.BLOCK_SIZE)) % self.SCREEN_WIDTH
        new_y = (cur.y + (y * self.BLOCK_SIZE)) % self.SCREEN_HEIGHT
        
        new_position = Snake.Position(new_x, new_y)
        
        if new_position == self.food_position:
            self.eat_food()
        
        if len(self.positions) > 2 and any(p == new_position for p in self.positions[2:]):
            self.reset()
        else:
            self.positions.insert(0, new_position)
            if len(self.positions) > self.length:
                self.positions.pop()
    
    def random_food_position(self):
        while True:
            x = self.random.randint(0, self.SCREEN_WIDTH // self.BLOCK_SIZE) * self.BLOCK_SIZE
            y = self.random.randint(0, self.SCREEN_HEIGHT // self.BLOCK_SIZE) * self.BLOCK_SIZE
            new_food = Snake.Position(x, y)
            if new_food not in self.positions:
                self.food_position = new_food
                break
    
    def reset(self):
        self.length = 1
        self.positions = [Snake.Position(self.SCREEN_WIDTH // 2, self.SCREEN_HEIGHT // 2)]
        self.score = 0
        self.random_food_position()
    
    def eat_food(self):
        self.length += 1
        self.score += 100
        self.random_food_position()
    
    def get_length(self):
        return self.length
    
    def get_positions(self):
        return self.positions
    
    def get_score(self):
        return self.score
    
    def get_food_position(self):
        return self.food_position

File: data/derived_cot/rq1_traces/translation/glm5.1/java_to_cpp/code_output/Snake.cpp
#ifndef SNAKE_H
#define SNAKE_H

#include <vector>
#include <random>
#include <cstddef>

class Snake {
public:
    class Position {
    private:
        int x;
        int y;

    public:
        Position(int x, int y) : x(x), y(y) {}

        int getX() const { return x; }
        int getY() const { return y; }

        bool equals(const Position& o) const {
            return x == o.x && y == o.y;
        }

        bool operator==(const Position& o) const {
            return x == o.x && y == o.y;
        }
    };

private:
    int length;
    const int SCREEN_WIDTH;
    const int SCREEN_HEIGHT;
    const int BLOCK_SIZE;
    std::vector<Position> positions;
    int score;
    Position foodPosition;
    std::mt19937 random;

    // Java-style modulo: always returns non-negative result
    static int javaMod(int a, int b) {
        int r = a % b;
        return (r < 0) ? r + b : r;
    }

    bool containsPosition(const Position& pos) const {
        for (const auto& p : positions) {
            if (p.equals(pos)) return true;
        }
        return false;
    }

public:
    Snake(int SCREEN_WIDTH, int SCREEN_HEIGHT, int BLOCK_SIZE, Position foodPosition)
        : length(1),
          SCREEN_WIDTH(SCREEN_WIDTH),
          SCREEN_HEIGHT(SCREEN_HEIGHT),
          BLOCK_SIZE(BLOCK_SIZE),
          score(0),
          foodPosition(foodPosition),
          random(std::random_device{}())
    {
        positions.push_back(Position(SCREEN_WIDTH / 2, SCREEN_HEIGHT / 2));
    }

    void move(Position direction) {
        Position cur = positions[0];
        int x = direction.getX();
        int y = direction.getY();

        int newX = javaMod(cur.getX() + (x * BLOCK_SIZE), SCREEN_WIDTH);
        int newY = javaMod(cur.getY() + (y * BLOCK_SIZE), SCREEN_HEIGHT);

        Position newPosition(newX, newY);

        if (newPosition.equals(foodPosition)) {
            eatFood();
        }

        // Check if newPosition is in positions[2..end)
        bool collision = false;
        if (positions.size() > 2) {
            for (size_t i = 2; i < positions.size(); i++) {
                if (positions[i].equals(newPosition)) {
                    collision = true;
                    break;
                }
            }
        }

        if (collision) {
            reset();
        } else {
            positions.insert(positions.begin(), newPosition);
            if (positions.size() > static_cast<size_t>(length)) {
                positions.pop_back();
            }
        }
    }

    void randomFoodPosition() {
        std::uniform_int_distribution<int> distX(0, SCREEN_WIDTH / BLOCK_SIZE - 1);
        std::uniform_int_distribution<int> distY(0, SCREEN_HEIGHT / BLOCK_SIZE - 1);
        do {
            int x = distX(random) * BLOCK_SIZE;
            int y = distY(random) * BLOCK_SIZE;
            foodPosition = Position(x, y);
        } while (containsPosition(foodPosition));
    }

    void reset() {
        length = 1;
        positions.clear();
        positions.push_back(Position(SCREEN_WIDTH / 2, SCREEN_HEIGHT / 2));
        score = 0;
        randomFoodPosition();
    }

    void eatFood() {
        length++;
        score += 100;
        randomFoodPosition();
    }

    int getLength() const { return length; }

    std::vector<Position>& getPositions() { return positions; }

    int getScore() const { return score; }

    Position getFoodPosition() const { return foodPosition; }
};

#endif